src/psyclone/domain/lfric/kernel_interface.py

```diff
@@ -0,0 +1,841 @@
+    BasisFunctionQrFaceDataSymbol, BasisFunctionQrEdgeDataSymbol, \
+    DiffBasisFunctionQrXyozDataSymbol, DiffBasisFunctionQrFaceDataSymbol, \
+    DiffBasisFunctionQrEdgeDataSymbol, \
+    NumberOfQrPointsInHorizontalDataSymbol, \
+    NumberOfQrPointsInVerticalDataSymbol, NumberOfQrPointsDataSymbol, \
+    QrWeightsInHorizontalDataSymbol, QrWeightsInVerticalDataSymbol, \
+    NumberOfFacesDataSymbol, QrWeightsDataSymbol, NumberOfEdgesDataSymbol
+from psyclone.psyir.symbols import SymbolTable, ArgumentInterface
+from psyclone.psyir.frontend.fparser2 import INTENT_MAPPING
+from psyclone.errors import InternalError, GenerationError
+from psyclone.domain.lfric import FunctionSpace
+
+
+class KernelInterface(ArgOrdering):
+    '''Create the kernel arguments for the supplied kernel as specified by
+    the associated kernel metadata and the kernel ordering rules
+    encoded in the ArgOrdering base class as method callbacks.
+
+    A PSyIR symbol table is created containing appropriate LFRic PSyIR
+    symbols to specify the arguments. If an argument is an array with
+    one or more dimension sizes specified by another argument, then
+    the associated array symbol is created so that it references the
+    appropriate symbol.
+
+    Related arguments - e.g. a field has an associated dofmap - are
+    not directly connected, they must be inferred from the function
+    space names. It is not yet clear whether this would be useful or
+    not.
+
+    TBD: This class should replace the current kernel stub generation
+    code when all of its methods are implemented, see issue #928.
+
+    :param kern: the kernel that requires its kernel arguments to be \
+        created.
+    :type kern: :py:class:`psyclone.dynamo0p3.DynKern`
+
+    '''
+    def __init__(self, kern):
+        ArgOrdering.__init__(self, kern)
+        self._read_access = ArgumentInterface(ArgumentInterface.Access.READ)
+        self._symbol_table = SymbolTable()
+        self._arglist = []
+
+    def generate(self, var_accesses=None):
+        '''Call the generate base class then add the argument list as it can't
+        be appended as we go along.
+
+        :param var_accesses: an unused optional argument that stores \
+            information about variable accesses.
+        :type var_accesses: :\
+            py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        '''
+        super(KernelInterface, self).generate()
+        # Set the argument list for the symbol table. This is done at
+        # the end after incrementally adding symbols to the _arglist
+        # list, as it is not possible to add symbols to the symbol
+        # table argument list.
+        self._symbol_table.specify_argument_list(self._arglist)
+
+    def cell_position(self, var_accesses=None):
+        '''Create an LFRic cell-position object and add it to the symbol table
+        and argument list.
+
+        :param var_accesses: an unused optional argument that stores \
```

Comment:
So why is it needed? We are capturing the required information in the symbol table so could provide information based on that if necessary. I guess my question comes from not understanding what var_accesses is used for (dependence analysis presumably?), but why would we need this for KernelInterface?

Reply:
Yes, it's for dependence analysis. If `KernelInterface` is used to describe a kernel call then it captures how the symbols passed to it are accessed (r, rw etc.). This information needs to be included in `var_accesses` for the dependence analysis to work. I think that's right. @hiker can correct me if I'm wrong.

Reply:
Doing this seems wrong to me. We have captured the access information in the symbol table so I think the dependence analysis should use this, rather than this class having to populate another list with redundant information. Please tell me if I'm talking out of turn @hiker. If @hiker is not against this approach I will create another issue and reference it in the class.

Reply:
I agree that we have the necessary information in the symbol table and that that should be our one truth. However, the dependence analysis does not yet use Symbols - see #845. If @hiker agrees we can simply add a TODO and refer to that issue. Especially since the functionality added by this PR is not yet used for actual code generation. (Once it is, it will probably/should break a number of tests related to dependence analysis.)